kong/db/dao/plugins.lua

```diff
@@ -157,6 +165,14 @@ local function load_plugin_handler(plugin)
     return nil, plugin .. " plugin is enabled but not installed;\n" .. handler
   end
 
+  if implements(handler, "response") and
+      (implements(handler, "header_filter") or implements(handler, "body_filter"))
+  then
+    return nil, fmt(
+      "Plugin %q can't be loaded because it implements both `response` " ..
+      "and `header_filter` or `body_filter` methods.\n", plugin)
+  end
```

Comment:
Not technical, but the reasoning was like:

- there should be no reason to use both.
- buffering modifies the `x_filter` methods behaviour (mostly a superset, but still different)
- even though it's documented, it's not too evident that adding `response` turns on buffering.
- if an existing plugin has `x_filter` phases, adding `response` _should_ require a rewriting, since existing behaviour would be affected.

so, it's just about making things evident by forcing the developer to rethink.  of course, it's true that one plugin can turn buffering on and affect other plugins, but that's probably uncommon.

Reply:
> if an existing plugin has x_filter phases, adding response should require a rewriting, since existing behaviour would be affected.

If I take any of our plugin that implements `x_filter` and add empty `response` function to them, how the need to be rewritten? I understand that it turns on buffering, but that doesn't change anything on `_filter` phases (yes, it is possible to read body on header filter, but that is mostly about it, and it should not affect any existing plugin).

Reply:
My reasoning is still the same. If any of our plugins implement any phase, that doesn't prevent them implementing any other phase. This is the first that introduces such limitation, and I don't still see a reason.

Let's say plugin A implements response and header_filter and has priority of 1, Plugin B implements response phase and has priority of 100.

1. Plugin A runs response phase
2. Plugin B runs response phase and sets response header X-Dog: Pluto
3. Plugin A wants to run header_filter and set X-Dog: Bulivar

1. and 2. works, but 3. does not, and I really don't understand why. What does the Plugin writer A need to rethink?

Reply:
I understand that the ADR said:
> The presence of the response callback in a plugin enables "buffered mode" which enables buffered proxying
> and disables the filter callbacks when loading the plugin, we can error if there are both response and filter
> callbacks with a nice explanation

But my feedback on that was not available back then. Now that I have given it, with example, could we at least rethink about it? @hishamhm?